swarm/pss/pss_test.go

```diff
@@ -475,6 +482,196 @@ func testAsymSend(t *testing.T) {
+	}
+	net := simulations.NewNetwork(adapter, &simulations.NetworkConfig{
+		ID: "0",
+	})
+	defer net.Shutdown()
+
+	f, err := os.Open(fmt.Sprintf("testdata/snapshot_%s.json", paramstring[1]))
+	if err != nil {
+		t.Fatal(err)
+	}
+	jsonbyte, err := ioutil.ReadAll(f)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var snap simulations.Snapshot
+	err = json.Unmarshal(jsonbyte, &snap)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = net.Load(&snap)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	triggerChecks := func(trigger chan discover.NodeID, id discover.NodeID, rpcclient *rpc.Client) error {
+		msgC := make(chan APIMsg)
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+		defer cancel()
+		sub, err := rpcclient.Subscribe(ctx, "pss", msgC, "receive", hextopic)
+		if err != nil {
+			t.Fatal(err)
+		}
+		go func() {
+			defer sub.Unsubscribe()
+			for {
+				select {
+				case recvmsg := <-msgC:
+					idx, _ := binary.Uvarint(recvmsg.Msg)
+					if recvmsgs[idx] == false {
+						log.Debug("msg recv", "idx", idx, "id", id)
+						recvmsgs[idx] = true
+						trigger <- id
+					}
+				case <-sub.Err():
+					return
+				}
+			}
+		}()
+		return nil
+	}
+
+	for i, nod := range net.GetNodes() {
+		nodes[i] = nod.ID()
+		rpcs[nodes[i]], err = nod.Client()
+		if err != nil {
+			t.Fatal(err)
+		}
+		pubkey := make([]byte, 65)
+		pubkeys[nod.ID()] = make([]byte, 65)
+		err = rpcs[nodes[i]].Call(&pubkey, "pss_getPublicKey")
+		if err != nil {
+			t.Fatal(err)
+		}
+		copy(pubkeys[nod.ID()], pubkey)
+		addr := make([]byte, 32)
+		err = rpcs[nodes[i]].Call(&addr, "pss_baseAddr")
+		if err != nil {
+			t.Fatal(err)
+		}
+		bzzaddrs[nodes[i]] = addr[:addrsize]
+		err = triggerChecks(trigger, nodes[i], rpcs[nodes[i]])
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	messagedelayhigh := 0
+	for i := 0; i < int(msgcount); i++ {
+		sendnodeidx := rand.Intn(int(nodecount))
+		recvnodeidx := rand.Intn(int(nodecount - 1))
+		if recvnodeidx >= sendnodeidx {
+			recvnodeidx++
+		}
+		nodemsgcount[nodes[recvnodeidx]]++
+		sentmsgs[i] = make([]byte, 8)
+		c := binary.PutUvarint(sentmsgs[i], uint64(i))
+		if c == 0 {
+			t.Fatal("0 byte message")
+		}
+		err = rpcs[nodes[sendnodeidx]].Call(nil, "pss_setPeerPublicKey", pubkeys[nodes[recvnodeidx]], hextopic, bzzaddrs[nodes[recvnodeidx]])
+		if err != nil {
+			t.Fatal(err)
+		}
+		err = rpcs[nodes[recvnodeidx]].Call(nil, "pss_setPeerPublicKey", pubkeys[nodes[sendnodeidx]], hextopic, bzzaddrs[nodes[sendnodeidx]])
+		if err != nil {
+			t.Fatal(err)
+		}
+		messagedelay := rand.Intn(messagedelayvarianceusec)
+		if messagedelay > messagedelayhigh {
+			messagedelayhigh = messagedelay
+		}
+		messagedelayduration, err := time.ParseDuration(fmt.Sprintf("%dus", messagedelay))
+		if err != nil {
+			t.Fatal(err)
+		}
+		go func(msg []byte) {
+			time.Sleep(messagedelayduration)
+			err = rpcs[nodes[sendnodeidx]].Call(nil, "pss_sendAsym", common.ToHex(pubkeys[nodes[recvnodeidx]]), hextopic, msg)
```

Comment:
All variables declared in the loop body need to be passed into the goroutine function (so `messageDelayDuration`, `sendNodeIndex` and `recvNodeIndex`).

Also, since `rpcs` and `pubkeys` are maps, this will likely fail under the race detector as many goroutines will be accessing the map concurrently. Consider using `sync.Map`.

Reply:
only the ones that change in the iteration.

also: the map lookups can be done before outside of the go routing

Reply:
> only the ones that change in the iteration.

Ah yes, what I said is bogus :grin:.

> the map lookups can be done before outside of the go routing

Indeed.